### PR TITLE
Optimize `table.init` instruction and instantiation

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit optimizes table initialization as part of instance
instantiation and also applies the same optimization to the `table.init`
instruction. One part of this commit is to remove some preexisting
duplication between instance instantiation and the `table.init`
instruction itself, after this the actual implementation of `table.init`
is optimized to effectively have fewer bounds checks in fewer places and
have a much tighter loop for instantiation.

A big fallout from this change is that memory/table initializer offsets
are now stored as `u32` instead of `usize` to remove a few casts in a
few places. This ended up requiring moving some overflow checks that
happened in parsing to later in code itself because otherwise the wrong
spec test errors are emitted during testing. I've tried to trace where
these can possibly overflow but I think that I managed to get
everything.

In a local synthetic test where an empty module with a single 80,000
element initializer this improves total instantiation time by 4x (562us
=> 141us)

### CODE REVIEW

File: crates/runtime/src/instance.rs [hunk 1]
```diff
@@ -573,11 +573,11 @@ impl Instance {
...
         }
 
-        Some(unsafe { &*self.anyfunc_ptr(index) })
+        unsafe { Some(&*self.vmctx_plus_offset(self.offsets.vmctx_anyfunc(index))) }
     }
 
-    unsafe fn anyfunc_ptr(&self, index: FuncIndex) -> *mut VMCallerCheckedAnyfunc {
-        self.vmctx_plus_offset(self.offsets.vmctx_anyfunc(index))
+    unsafe fn anyfunc_base(&self) -> *mut VMCallerCheckedAnyfunc {
+        self.vmctx_plus_offset(self.offsets.vmctx_anyfuncs_begin())
```
Comment (fitzgen) [inline crates/runtime/src/instance.rs:580]: Why this change? The `anyfunc_ptr` call should be inlined just fine, unless I'm missing something. This just makes things more open coded, which isn't the direction I'd generally move in.

File: crates/runtime/src/table.rs [hunk 2]
```diff
@@ -212,6 +212,33 @@ impl Table {
...
+            let elements = match elements
+                .get_mut(usize::try_from(dst).unwrap()..)
+                .and_then(|s| s.get_mut(..items.len()))
+            {
+                Some(elements) => elements,
+                None => return Err(Trap::wasm(ir::TrapCode::TableOutOfBounds)),
+            };
+
+            for (item, slot) in items.zip(elements) {
+                Self::set_raw(ty, slot, item.into())
```
Comment (fitzgen) [inline crates/runtime/src/table.rs:236]: Does this get boiled down into a `memcpy` by LLVM? If so, we can close #983. If not, would we be able to do that here?

File: crates/runtime/src/table.rs [hunk 3]
```diff
@@ -212,6 +212,33 @@ impl Table {
...
+    /// Fill `table[dst..]` with values from `items`
+    ///
+    /// Returns a trap error on out-of-bounds accesses.
+    pub fn init_funcs(
+        &self,
+        dst: u32,
+        items: impl ExactSizeIterator<Item = *mut VMCallerCheckedAnyfunc>,
+    ) -> Result<(), Trap> {
+        let ty = TableElementType::Func;
+        assert!(ty == TableElementType::Func);
```
Comment (fitzgen) [inline crates/runtime/src/table.rs:224]: This seems very tautological as written; is this supposed to be checking `self.element_type`?

File: crates/runtime/src/table.rs (hunk 2 above)
Comment (fitzgen) [reply crates/runtime/src/table.rs:236]: It seems like this has just one call site, and I don't think anything is preventing us from doing a `memcpy` in theory here.

**Review by fitzgen — APPROVED** [review]
r=me with comments below addressed

File: crates/runtime/src/instance.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/runtime/src/instance.rs:580]: This mostly changed because it was previously used in a loop but the pattern it was using was a bit slower than necessary. The main slowdown is that `vmctx_anyfunc` performed a multiplication instruction with the size of a pointer, where both the size and the index were not constant, resulting in bad codegen. By using the native types here it makes pointer arithmetic much speedier since everything is known at compile time.

Once I removed one main use of the function in the table initialization bits I figured I might as well go ahead and remove the function and speed up other callsites as well.

I do agree it's a bit more open coded, but then again everything about `VMContext` feels sort of inherently open-coded

File: crates/runtime/src/table.rs (hunk 2 above)
Comment (alexcrichton) [reply crates/runtime/src/table.rs:236]: Oh this should actually probably close that issue. We don't actually have anything to `memcpy` though because our element segments are stored as lists of function indices, but the tables themselves are `*mut VMCallerCheckedAnyfunc`. This means that there's a bit of a mismatch where unless we allocate table segments per-instance to `memcpy` from (which I dont think we should do) then there's actually nothing to memcpy. 

Otherwise, though, I just used `set_raw` here to be the same as the other bits and pieces of this module

File: crates/runtime/src/table.rs (hunk 3 above)
Comment (alexcrichton) [reply crates/runtime/src/table.rs:224]: Oops, yes!

File: crates/runtime/src/table.rs (hunk 2 above)
Comment (alexcrichton) [reply crates/runtime/src/table.rs]: I went ahead though and made it a bit more raw to help streamline this a tiny bit more

File: crates/runtime/src/instance.rs (hunk 1 above)
Comment (fitzgen) [reply crates/runtime/src/instance.rs:580]: Huh, I would really have expected `anyfunc_ptr` to be inlined and then for inst combine to see that one of the operands of the multiplication was a constant and go to town from there.
